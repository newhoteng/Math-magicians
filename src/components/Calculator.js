import React from 'react';
import PropTypes from 'prop-types';

// DigitButton component with prop validation
function DigitButton({ digit }) {
  return (
    <button type="button" className="digit-button">{ digit }</button>
  );
}
DigitButton.propTypes = {
  digit: PropTypes.oneOfType(PropTypes.string, PropTypes.number).isRequired,
};

// OpertorButton component with prop validation
function OperatorButton({ operator }) {
  return (
    <button type="button" className="operator-button">{ operator }</button>
  );
}
OperatorButton.propTypes = {
  operator: PropTypes.string.isRequired,
};

// Calculator App
function Calculator() {
  const Operators = ['รท', 'x', '-', '+', '='];
  const Digits = ['AC', '+ / -', '%', 7, 8, 9, 4, 5, 6, 1, 2, 3, 0, '.'];

  // Create button elements from Operators array using OperatorButton component
  const operatorList = Operators.map((operator) => (
    <OperatorButton key={operator} operator={operator} />
  ));

  // Create button elements from Digits array using DigitButton component
  const digitList = Digits.map((digit) => (
    <DigitButton key={digit} digit={digit} />
  ));

  return (
    <div className="calculator-wrapper">
      <div className="display-panel">0</div>
      <div className="digit-wrapper">
        {digitList}
      </div>
      <div className="operator-wrapper">
        {operatorList}
      </div>
    </div>
  );
}

export default Calculator;
