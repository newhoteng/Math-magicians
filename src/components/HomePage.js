function HomePage() {
  return (
    <div>
      <p>
        Lorem ipsum dolor sit amet consectetur adipisicing elit.
        Maxime reprehenderit eaque esse obcaecati numquam impedit culpa? Quae numquam deserunt quas!
        Ad animi inventore laborum debitis non consequuntur exercitationem dolor quaerat?
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Corporis impedit vel accusantium
        iste dicta possimus, nisi, doloribus vitae, cum autem tenetur!
        Ea quisquam nam fugit necessitatibus quae quidem, sit minima.
      </p>
      <p>
        Lorem ipsum dolor sit amet consectetur adipisicing elit.
        Maxime reprehenderit eaque esse obcaecati numquam impedit culpa? Quae numquam deserunt quas!
        Ad animi inventore laborum debitis non consequuntur exercitationem dolor quaerat?
        Lorem ipsum dolor sit amet consectetur adipisicing elit.
        Corporis impedit vel accusantium iste dicta possimus, nisi, doloribus vitae,
        cum autem tenetur! Ea quisquam nam fugit necessitatibus quae quidem, sit minima.
      </p>
    </div>
  );
}

export default HomePage;
