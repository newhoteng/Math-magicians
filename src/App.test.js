import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Calculate from './logic/calculate';
import Operate from './logic/operate';
import HomePage from './components/HomePage';
import Calculator from './components/Calculator';
import Quote from './components/Quotes';

describe('calculations', () => {
  test('calculate 2 x 2 + 16', () => {
    // Arrange
    const obj = { total: 0, next: null, operation: null };

    // Act
    const clickOnFirst2 = Calculate(obj, '2');
    const clickOnx = Calculate(clickOnFirst2, 'x');
    const clickOnSecond2 = Calculate(clickOnx, '2');
    const clickOnX2 = Calculate(clickOnSecond2, '+');
    const clickOn16 = Calculate(clickOnX2, '16');
    const clickOnEqualButton = Calculate(clickOn16, '=');

    // Assert
    expect(clickOnEqualButton.total).toBe('20');
  });

  test('calculate 3 + 12 ÷ 3', () => {
    // Arrange
    const obj = { total: 0, next: null, operation: null };

    // Act
    const clickOnFirst3 = Calculate(obj, '3');
    const clickOnplus = Calculate(clickOnFirst3, '+');
    const clickOn12 = Calculate(clickOnplus, '12');
    const clickOndiv = Calculate(clickOn12, '÷');
    const clickOn3 = Calculate(clickOndiv, '3');
    const clickOnEqualButton = Calculate(clickOn3, '=');

    // Assert
    expect(clickOnEqualButton.total).toBe('5');
  });

  test('calculate 1 ÷ 4', () => {
    // Arrange
    const obj = { total: 0, next: null, operation: null };

    // Act
    const clickOnFirstTwo = Calculate(obj, '1');
    const clickOnX = Calculate(clickOnFirstTwo, '÷');
    const clickOnSecondTwo = Calculate(clickOnX, '4');
    const clickOnEqualButton = Calculate(clickOnSecondTwo, '=');

    // Assert
    expect(clickOnEqualButton.total).toBe('0.25');
  });

  test('calculate 1 ÷ 4 AC', () => {
    // Arrange
    const obj = { total: 0, next: null, operation: null };

    // Act
    const clickOnFirstTwo = Calculate(obj, '1');
    const clickOnX = Calculate(clickOnFirstTwo, '÷');
    const clickOnSecondTwo = Calculate(clickOnX, '4');
    const clickOnEqualButton = Calculate(clickOnSecondTwo, 'AC');

    // Assert
    expect(clickOnEqualButton.total).toBe('0');
  });

  test('calculate 8 + / -', () => {
    // Arrange
    const obj = { total: 0, next: null, operation: null };

    // Act
    const clickOnFirstTwo = Calculate(obj, '8');
    const clickOnX = Calculate(clickOnFirstTwo, '+ / -');

    // Assert
    expect(clickOnX.next).toBe('-8');
  });
});

describe('operations', () => {
  test('operation 16 x 2', () => {
    // Arrange
    const items = [16, 2, 'x'];

    // Act
    const result = Operate(...items);

    // Assert
    expect(result).toBe('32');
  });

  test('operation 56 ÷ 0.5', () => {
    // Arrange
    const items = [56, 0.5, '÷'];

    // Act
    const result = Operate(...items);

    // Assert
    expect(result).toBe('112');
  });

  test('operation 1021 - 30', () => {
    // Arrange
    const items = [1021, 30, '-'];

    // Act
    const result = Operate(...items);

    // Assert
    expect(result).toBe('991');
  });
  //  start changes
  test('operation 102 + 30', () => {
    // Arrange
    const items = [102, 30, '+'];

    // Act
    const result = Operate(...items);

    // Assert
    expect(result).toBe('132');
  });

  test('operation 46 ÷ 0', () => {
    // Arrange
    const items = [46, 0, '÷'];

    // Act
    const result = Operate(...items);

    // Assert
    expect(result).toBe('Can\'t divide by 0.');
  });

  test('operation 46 % 0', () => {
    // Arrange
    const items = [46, 0, '%'];

    // Act
    const result = Operate(...items);

    // Assert
    expect(result).toBe('Can\'t find modulo as can\'t divide by 0.');
  });

  test('operation 4 $ 6', () => {
    // Arrange
    const items = [4, 6, '^'];
    const expectedErrorMessage = /Unknown operation '\^'/;

    // Act and Assert
    expect(() => Operate(...items)).toThrowError(expectedErrorMessage);
  });
});

describe('App snapshots', () => {
  it('renders correctly', () => {
    const homepage = renderer
      .create(<HomePage page="/">Math-magician</HomePage>)
      .toJSON();
    expect(homepage).toMatchSnapshot();
  });

  it('renders correctly', () => {
    const calculatorpage = renderer
      .create(<Calculator page="/calculator">Math-magician/calculator</Calculator>)
      .toJSON();
    expect(calculatorpage).toMatchSnapshot();
  });

  it('renders correctly', () => {
    const quotepage = renderer
      .create(<Quote page="/quote">Math-magician/quote</Quote>)
      .toJSON();
    expect(quotepage).toMatchSnapshot();
  });
});

describe('App simulate user interaction', () => {
  it('renders Homepage component', () => {
    render(<HomePage />);
  });

  it('renders Calculator component', () => {
    render(<Calculator />);
  });

  it('renders Quote component', () => {
    render(<Quote />);
  });
});
