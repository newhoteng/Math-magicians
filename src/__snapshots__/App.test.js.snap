// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshots renders correctly 1`] = `
<div
  className="homediv"
>
  <h1>
    Welcome to our page!
  </h1>
  <p
    className="homep"
  >
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Maxime reprehenderit eaque esse obcaecati numquam impedit culpa? Quae numquam deserunt quas! Ad animi inventore laborum debitis non consequuntur exercitationem dolor quaerat? Lorem ipsum dolor sit amet consectetur adipisicing elit. Corporis impedit vel accusantium iste dicta possimus, nisi, doloribus vitae, cum autem tenetur! Ea quisquam nam fugit necessitatibus quae quidem, sit minima.
  </p>
  <p
    className="homep"
  >
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Maxime reprehenderit eaque esse obcaecati numquam impedit culpa? Quae numquam deserunt quas! Ad animi inventore laborum debitis non consequuntur exercitationem dolor quaerat? Lorem ipsum dolor sit amet consectetur adipisicing elit. Corporis impedit vel accusantium iste dicta possimus, nisi, doloribus vitae, cum autem tenetur! Ea quisquam nam fugit necessitatibus quae quidem, sit minima.
  </p>
</div>
`;

exports[`App snapshots renders correctly 2`] = `
<div
  className="calculator-div"
>
  <h1>
    Let's do some math!
  </h1>
  <div
    className="calculator-wrapper"
  >
    <div
      className="display-panel"
    >
      0
       
       
    </div>
    <div
      className="digit-wrapper"
    >
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        + / -
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        %
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        className="digit-button"
        onClick={[Function]}
        type="button"
      >
        .
      </button>
    </div>
    <div
      className="operator-wrapper"
    >
      <button
        className="operator-button"
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
      <button
        className="operator-button"
        onClick={[Function]}
        type="button"
      >
        x
      </button>
      <button
        className="operator-button"
        onClick={[Function]}
        type="button"
      >
        -
      </button>
      <button
        className="operator-button"
        onClick={[Function]}
        type="button"
      >
        +
      </button>
      <button
        className="operator-button"
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;

exports[`App snapshots renders correctly 3`] = `
<div
  className="quote-div"
>
  <blockquote
    className="quote-wrapper"
  >
    <p
      className="quote"
    >
      Loading...
    </p>
  </blockquote>
</div>
`;
